src/mon/MonMap.cc

```diff
@@ -823,9 +844,26 @@ int MonMap::init_with_dns_srv(CephContext* cct,
   }
 }
 
+void MonMap::update_cct(CephContext *cct)
+{
+  std::vector<entity_addrvec_t> mon_addrs;
+  for (auto& i : mon_info) {
+    mon_addrs.push_back(i.second.public_addrs);
+  }
+
+  cct->set_mon_addrs(mon_addrs);
+}
+
 int MonMap::build_initial(CephContext *cct, bool for_mkfs, ostream& errout)
 {
   const auto& conf = cct->_conf;
+
+  // cct?
+  auto addrs = cct->get_mon_addrs();
```

Comment:
what's the use case of this?

Reply:
In a long running daemon/process (like ceph-mgr), where the initial daemon is bootstrapped with mon addrs from a config or CLI arguments, if the mon addresses change over time, future instantiations of MonClient and hence MonMap will fail to find any of the currently valid mon's.

For example libcephfs in ceph-mgr instantiates a MonClient based on fs mgr requests, and such instantiations relied on mon config as present during mgr startup, which may now be stale.

Reply:
i see. without this change, we don't have a machinery passing persisted monmap retrieved from monitor after the a MonClient gets authenticated to the newly created librados client in a long running daemon. and there are use cases where we could create **brand new** MonClient instances multiple times along the way for librados clients. because we don't pass the latest MonClient or the monmap to librados clients when instantiating them or when connecting them to the cluster. so, even if we have access to the latest monmap, those hosted librados clients do not have access to them, and they might fail to connect to the cluster because all the monitors have been migrated to new addresses.

i think this reveals an issue. the daemons surely have access to the latest monmap because they are up and running across the monitor migrations. but let's consider a cluster with 3 ceph-mon and 1 ceph-mgr:

1. ceph-mgr.x started. and it is the active mgr in the cluster.
1. mon.{a, b, c} migrated to new hosts one after another
1. we disable and enable some mgr module, ceph-mgr.x kills itself, and restarts for properly teardown and setup of the involved mgr module(s).
1. ceph-mgr.x is like newborn baby. and it has no idea of the new hosts of mon.a, mon.b or mon.c. so it is not able to connect to anyone of them.

so i think it's more about how to keep the minimal settings about the cluster so the daemons and clients can talk to their cluster and the monitors are relocated. and it boils down to how clients of librados initialize their initial monmap -- they use `MonMap::build_initial()`. see also https://tracker.ceph.com/issues/46645#note-1 .

and there are two cases:

1. clients hosted by a long running process which happen to have the monmap containing reachable monitor before it restarts. so we have couple options here:
  - stash the latest monmap in the shared place, and instruct monmap to use it as an alternative for the initial monmap along with the settings in `ceph.conf`. as discussed in https://tracker.ceph.com/issues/46645#note-1 .
  - use methods like `conf_set()` to inform it the latest `mon_host` before connecting it to cluster.
2. librados clients and other daemons (OSD for instance) using MonClient started by a new process. they really depends on `ceph.conf` with at least one reachable monitor.

so i think the crux of the problem is: how to populate the minimal cluster settings to a running process and to a new process.

it seems we already have an answer for the 2nd question: to push an updated copy of `ceph.conf` to the host. but i think we also have an answer for the 1st question already: 1) to push a new copy of `ceph.conf` to the host, and also 2) `ceph config set`.  so we can cater the needs of the long running process and the MonClients instantiated by it, and also the needs of the long running process after it restarts. i think it should be documented as part of https://ceph.readthedocs.io/en/latest/rados/deployment/ceph-deploy-mon/ .

Reply:
> it seems we already have an answer for the 2nd question: to push an updated copy of `ceph.conf` to the host. but i think we also have an answer for the 1st question already: 1) to push a new copy of `ceph.conf` to the host,

(1) isn't feasible for containerized setups I believe. Plus you'd need to tell the daemon to re-read its config.

> and also 2) `ceph config set`. so we can cater the needs of the long running process and the MonClients instantiated by it, and also the needs of the long running process after it restarts. i think it should be documented as part of https://ceph.readthedocs.io/en/latest/rados/deployment/ceph-deploy-mon/ .

You can't update `mon host` through `ceph config set` due to the `NO_MON_UPDATE` flag. We could change that but I'd worry about unintended consequences.

Reply:
> > it seems we already have an answer for the 2nd question: to push an updated copy of `ceph.conf` to the host. but i think we also have an answer for the 1st question already: 1) to push a new copy of `ceph.conf` to the host,
> 
> (1) isn't feasible for containerized setups I believe. 

i am not sure why it's not feasible. we always need to push settings to push settings to a host even if it resides in a container.

> Plus you'd need to tell the daemon to re-read its config.

yeah, need to call `parse_config_files()` for re-reading `ceph.conf`.

> 
> > and also 2) `ceph config set`. so we can cater the needs of the long running process and the MonClients instantiated by it, and also the needs of the long running process after it restarts. i think it should be documented as part of https://ceph.readthedocs.io/en/latest/rados/deployment/ceph-deploy-mon/ .
> 
> You can't update `mon host` through `ceph config set` due to the `NO_MON_UPDATE` flag. We could change that but I'd worry about unintended consequences.

ack.